### DISCUSSION TITLE
Possible to collect all pytest tests inside python

### DISCUSSION THREAD

**Discussion by jscotka** [pytest-dev/pytest#8554] [Q&A] [answered]
Hi,
I wanted to generate metadata for another project based on pytest tests (https://github.com/jscotka/fmf_metadata/)
And I was not able to create own implementation of `pytest --collect-only ` in python

I need something like this pseudo code:
```
import pytest
import _pytest
def collect()
    conf = _pytest.get_config("path_to_config")
    sess = session.Session(conf)
    sess.collect()
    return sess.items
```
I've tried to do that (playing wih config, session. fscollector, pluginmanager and more) but finally I was not successful and I had to workaround it by using `autouse` fixture with `scope="session"`, but it needs to add this fixture into `contest.py` https://github.com/jscotka/fmf_metadata/blob/main/fmf_metadata/pytest_fixture.py

My idea is to have offline tool without modifying test code itself.

Is there any way to do it, could anyone help me to implement it ?

**Comment by hoefling** [comment, marked as answer]
I gave an [answer to a similar question](https://stackoverflow.com/a/62955267/2650249) a while ago: you can register a plugin that stores the items and run `pytest.main` with the plugin supplied. Example code:
```py
import pytest


class ItemsCollector:
    def pytest_collection_modifyitems(self, items):
        self.items = items[:]


def main():
    collector = ItemsCollector()
    pytest.main(['--collect-only'], plugins=[collector])
    print("collected items:", collector.items)
```
It would be interesting to know though whether a better approach exists.

**Reply by RonnyPfannschmidt** [reply]
this one is the best approach as of now

**Reply by jscotka** [reply]
@hoefling many thanks.

**Reply by nicoddemus** [reply]
A quick note: you probably also want to pass `-p no:terminal` to suppress all pytest output:

```python
pytest.main(['--collect-only', '-pno:terminal'], plugins=[collector])
```

**Reply by nicoddemus** [reply]
@jscotka could you mark @hoefling's answer as accepted? This helps others looking for the same solution in the future. Thanks!

**Reply by jscotka** [reply]
Hi @nicoddemus thanks a lot. This terminal supress suggestion is great. Yes I'll mark it as accepted.

**Comment by jscotka** [comment]
Hi,
finally I've found the solution for my question, I  don'tknow if this solution is optimal, but works as exoected for me. Maybe if anybody have time, could me help me to do it more optimal if possible.
 
```python
from _pytest import config
from _pytest.main import Session


def collect(opts):
    conf = config.get_config(opts, plugins=[])
    conf.parse(opts)
    conf._do_configure()
    sess = Session.from_config(config=conf)

    conf.hook.pytest_sessionstart(session=sess)
    conf.hook.pytest_collection(session=sess)
    conf.hook.pytest_collection_finish(session=sess)

    return sess.items
```

**Reply by RonnyPfannschmidt** [reply]
@hoefling provided a much better solution, yours is hooking into internals it shouldn't touch

the stuff you use is something i intend to break within the next 6 months

**Reply by jscotka** [reply]
@RonnyPfannschmidt 
Thank you very much. It is exactly point where I had fear to use this solution.
